api/pkg/apis/v1alpha1/managers/activations/activations-manager.go

```diff
@@ -192,32 +192,74 @@ func (t *ActivationsManager) ReportStatus(ctx context.Context, name string, name
 	defer observ_utils.CloseSpanWithError(span, &err)
 	lock.Lock()
 	defer lock.Unlock()
-	getRequest := states.GetRequest{
-		ID: name,
+
+	var activationState model.ActivationState
+	activationState, err = t.GetState(ctx, name, namespace)
+	if err != nil {
+		return err
+	}
+
+	current.UpdateTime = time.Now().Format(time.RFC3339) // TODO: is this correct? Shouldn't it be reported?
+	activationState.Status = &current
+
+	var entry states.StateEntry
+	entry.ID = activationState.ObjectMeta.Name
+	entry.Body = activationState
+
+	upsertRequest := states.UpsertRequest{
+		Value: entry,
 		Metadata: map[string]interface{}{
 			"version":   "v1",
 			"group":     model.WorkflowGroup,
 			"resource":  "activations",
-			"namespace": namespace,
+			"namespace": activationState.ObjectMeta.Namespace,
+			"kind":      "Activation",
+		},
+		Options: states.UpsertOption{
+			UpdateStateOnly: true,
 		},
 	}
-	var entry states.StateEntry
-	entry, err = t.StateProvider.Get(ctx, getRequest)
+	_, err = t.StateProvider.Upsert(ctx, upsertRequest)
 	if err != nil {
 		return err
 	}
+	return nil
+}
+
+func (t *ActivationsManager) ReportStageStatus(ctx context.Context, name string, namespace string, current model.ActivationStatus) error {
+	ctx, span := observability.StartSpan("Activations Manager", ctx, &map[string]string{
+		"method": "ReportStageStatus",
+	})
+	var err error = nil
+	defer observ_utils.CloseSpanWithError(span, &err)
+	lock.Lock()
+	defer lock.Unlock()
 
 	var activationState model.ActivationState
-	bytes, _ := json.Marshal(entry.Body)
-	err = json.Unmarshal(bytes, &activationState)
+	activationState, err = t.GetState(ctx, name, namespace)
 	if err != nil {
-		observ_utils.CloseSpanWithError(span, &err)
 		return err
 	}
 
 	current.UpdateTime = time.Now().Format(time.RFC3339) // TODO: is this correct? Shouldn't it be reported?
+	inputString, _ := json.Marshal(current.Inputs)
+	outputString, _ := json.Marshal(current.Outputs)
+	if current.Status == v1alpha2.Done {
+		current.History = append(activationState.Status.History, model.StageStatus{
```

Comment:
activationState.Status is a reference with omit empty. Better to do nil check on activationState.Status before appending.

Reply:
getActivationState has initialized status field if empty.

Reply:
As the internal sync, we need to reorganize the properties in ActivationStatus. Given we have stageHistory, then we don't need to keep the current stage status separately from the root of ActiviationStatus. Please change the PR to draft before you complete the optimization.